### ISSUE TITLE
Forget the past, embrace the filters

### ISSUE THREAD

**Issue by ichthus1604** [molnard/The-release-game-v2-0-4-1#12]
**📜 Preamble**

Transaction History Filters by @ichthus1604 

**📖 TL;DR - Too Long, Didn't Read**

No one needs to see their transactions from a year ago in the main UI.
Introduce filters in the transaction history by date / send or receive / labels / transaction type

**🌟 Specification, Scope, and Features**

Add an (expandable/collapsible?) filter panel on top of the transaction history view. This would affect only the list and not interact with anything else such as the privacy / balance tiles.

By default, it could show only the transactions from last 30 days and hide anything before that. This way, if a new transaction comes in, the filter would include it and it would show up in the list, maintaining the current behavior where newly processed transactions are highlighted.

If the filter is set by the user to the past, or any filter option which does not include the newly processed transaction, the UI notification should be smart enough to show some message like "click to show" and give the user the ability to intentionally click on that and have the filters reset to default.

**⚔️ Motivation**

Endlessly scrolling through the transaction history (in whatever form, even if it's not a datagrid like @soosr proposed [here](https://github.com/molnard/The-release-game-v2-0-4-1/issues/3)) makes no sense and is no good UX. If I'm searching for whatever happened in 2022 I should be able to search by dates, months or years and have that information readily presented and filtered out from the rest. 

It also introduces many performance considerations as we're right now loading the entire lifetime of every wallet (in a multi wallet UI) in order to show the transaction history.

**📜 Rationale**

[Recent findings](https://github.com/zkSNACKs/WalletWasabi/pull/11458) by @turbolay show that we could significantly reduce memory usage by simply making transactions (and the related support objects that we use to show them in the UI) use less memory. How about letting them use 0 (zero) bytes of RAM by simply not loading them unless requested by the user?
 
On top of that, loading the entire wallet history of transactions shouldn't be necessary to display the current wallet state. Balance and other "current snapshot" data could be stored in the Sqlite Db and updated in an event-driven fashion.

This also goes in line with #10 (Optimize wallet performance for mobile)

**🌌 Backward Compatibility**

We already achieved migration into Sqlite in the past, we can do it again, this time for transactions and wallet state data.

**📚 Reference Implementation (optional)**

Pretty much every fintech application in the world has some sort of filtering of transactions.

--------------------------------------------------
🧑‍🏫 Team involvement

VDG Team
Lead: @ichthus1604 
UI/UX Design: @soosr
UI Implementation: @ichthus1604 

Code Team:
I propose @kiminuo take the lead in the Code side of things as he's [already been working](https://github.com/zkSNACKs/WalletWasabi/pull/11421) in what's required to make this happen

💪 Subtasks

This project can be divided into 2 large milestones:

 - Stage One: A UI-only implementation that simply filters in-memory data 
 - Stage Two: A full implementation that goes all the way down to the SQL for filtering

🚩Stage one Subtasks

 - [ ] [Create UX/UI Designs](https://github.com/zkSNACKs/WalletWasabi/issues/11655): @soosr
     - Date filters
     - Direction filters (incoming / outgoing transaction)
     - Transaction Type filters (regular, coinjoin, self-transfer, etc)
     - Label filters
     - Amount filters? (below/above X amount)
     - Define if and how these filters interact with the Search Bar.
     - Define if and how the "New Transaction" notification interacts with the filters

 - [ ] [Implement UI changes](https://github.com/zkSNACKs/WalletWasabi/issues/11656): @ichthus1604 🕐 1 1/2 weeks (subject to Design changes above)
   - Add Filters bar with all of the above: 2 days
   - Make Transaction History grid honor the filters: 1 day
   - Remove dependencies from all other ViewModels to the Transaction History: 3 days (this is already WIP and is part of UI Decoupling)
   - Modify the Transaction Received notifications: 1 day
   - Implement the interaction between the notifications and the filters: 1 day

 - [ ] [Define filtering API Contract](https://github.com/zkSNACKs/WalletWasabi/issues/11657): @ichthus1604 🕐1 Day
   - Define the common API that will be shared by UI and Client code to retrieve filtered transaction history.
   
🚩 Stage Two Subtasks:

 - [ ] Code side changes
     - Persist Wallet Stats in SQL
     - Persist Transactions in SQL
     - Persist Wallet Coins in SQL
     - Adapt all Client code to the new reality (things aren't fully loaded in memory anymore and whatever needs to be computed must not assume that what's in memory is all that exists)
     - Add required caches
     - Create migration scripts to move transactions from files to SQL in order to avoid full resync
     - Add required APIs for UI to query using filters

 - [ ] UI changes @ichthus1604 🕐 3 days
     - Change in-memory filtering to use the new transaction query APIs: 1 day
     - Change in-memory computation of wallet stats to use the new APIs: 1 day
     - Testing and bug fixing: 1 day

--------------------------------------------
### Next Steps:

I had a conversation with @turbolay and @kiminuo regarding what's required in order to fully implement this end-to-end.

They are already working on that, but it will take a significant (still not estimated) amount of time due to the heavy refactoring needed to migrate to SQLite.

With this, my plan is to fully implement Stage 1 which can be done completely in the UI side.

Aside from this, I will create the basic API contract that the Client side code needs to provide for the UI for this feature.

**Comment by molnard** [comment]
Noble Author,

Your Quest has been selected for further progress! The devil is in the details - so you need to elaborate and show your quest is not evil. We only embrace quests that are leading us to the light! To prove that you will need to add the following Sections to the Quest description above: 

**⚠️ Sections to add to the Quest descriptions**

- Team involvement: [ui / code / integrations / marketing ] can be one or more. 
- Subtasks. Try to break it down as much as possible. Think about teams and jobs that can be done in parallel. Phrase it like a Todo list with elements. If something is not clear yet do not forget to add a task - "research this or that"... 

**ℹ️ Tips** 

- Try to break down the task to Teams and then to Subtasks. 
- Try to be as detailed as possible. 
- Do your best to estimate or to get an estimation from the right ppl. 
If required you can arrange a meeting and invite team leaders or ppl you believe could be involved or have something to add. 

🕐  Please make this the utmost importance - we do not have much time left!

**Comment by Pule08** [comment]
Estimated duration is ~4 weeks

**Comment by nopara73** [comment]
Requirements superseded by https://github.com/zkSNACKs/WalletWasabi/issues/11705
